### PR TITLE
Creating an extendable structure

### PR DESCRIPTION
**Pull request by adamlith**
Hi,

I have a usecase where I need to be able to extend this framework.
I saw that one of the issues on this repo (https://github.com/destructurama/attributed/issues/18) was regarding something similar.
If you like this refactoring then I can build the attributes requested in the issue.

### CODE REVIEW

**Comment by nblumhardt** [comment]
Hi Adam! Thanks for the PR.

This looks really nice - especially since it extends even to cases like `Destructure.ByTransforming(...)` elegantly 👍 

I'll scan through the changes quickly now, but might need to come back to it for a proper review when I have a spare moment. Thanks again!

File: src/Destructurama.Attributed/Attributed/BaseDestructuringAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+using Serilog.Events;
+
+namespace Destructurama.Attributed
+{
+    public abstract class BaseDestructuringAttribute : Attribute
```
[the remark this answers is not in the document]
Comment (adamlith) [reply src/Destructurama.Attributed/Attributed/BaseDestructuringAttribute.cs]: Definently, no problem.

File: src/Destructurama.Attributed/Attributed/BaseDestructuringAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+    {
+        internal abstract LogEventPropertyValue CreateLogEventPropertyValue(object value);
```
[the remark this answers is not in the document]
Comment (adamlith) [reply src/Destructurama.Attributed/Attributed/BaseDestructuringAttribute.cs]: Oh, I like these ideas, I will implement them right away.

[the remark this answers is not in the document]
Comment (adamlith) [reply src/Destructurama.Attributed/Attributed/BaseDestructuringAttribute.cs]: But is public really necessary, isnt protected enough?

**Comment by adamlith** [comment]
I also have several ideas for readability and performance improvements, but they should probably be in separate PRs, after this one is done.

**Comment by adamlith** [comment]
I'm also contemplating if we should make the ILogEventPropertyValueFactory a parameter of the TryCreateLogEventProperty method. It's not useful for any of the usecases I have, but the framework does become even more extendable that way. What do you think?

**Comment by adamlith** [comment]
Aw, sh*t, i couldnt stop my itchy fingers from performing the readability improvements. I keep telling myself that they will make it easier for you to review the end-result.

If you want me to roll-back this last commit and wait until this first part is done before i submit a new PR then just say the word and I'll do it :)

And if you're wondering why I replaced the recursive tail-call in the TryDestructure method with a while-loop it was to be able to reduce the number of calls to value.GetType()
(and I also think that for the average person a loop is easier to read/understand than recursion)

File: src/Destructurama.Attributed/Attributed/DestructuringAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace Destructurama.Attributed
+{
+    /// <summary>
+    /// The base class for all destructuring attributes.
+    /// Inherit from this class to extend this library.
+    /// </summary>
+    public abstract class DestructuringAttribute : Attribute
+    {
+        protected internal abstract bool TryCreateLogEventProperty(string name, object value, out LogEventProperty property);
```
Comment (nblumhardt) [inline src/Destructurama.Attributed/Attributed/DestructuringAttribute.cs]: :+1:

**Comment by nblumhardt** [comment]
Looking great!

> I'm also contemplating if we should make the ILogEventPropertyValueFactory a parameter of the TryCreateLogEventProperty method. It's not useful for any of the usecases I have, but the framework does become even more extendable that way. What do you think?

This probably does make sense - it'd be particularly useful for attributes on things like collection properties, where an implementation might need to recursively capture the elements in the collection using whatever policies have been defined for the logger :+1:

**Comment by nblumhardt** [comment]
Great! Merging now. Looking over this broad changeset makes me realise it's overdue for a good clean-up; I should have a chance this morning to dig into it.

Thanks for the great PR!

**Pull request merged**